File: 2022/july/29-07-2022/find-and-replace-pattern.py
class Solution:
    def findAndReplacePattern(self, words, pattern: str):
        result = []
        length = len(pattern)
        for word in words:
            tmp1 = {}
            tmp2 = {}
            if length == len(word):
                curr = 0
                while curr < length:
                    if pattern[curr] in tmp1:
                        if tmp1[pattern[curr]] != word[curr]:
                            break
                    else:
                        tmp1[pattern[curr]] = word[curr]

                    if word[curr] in tmp2:
                        if tmp2[word[curr]] != pattern[curr]:
                            break
                    else:
                        tmp2[word[curr]] = pattern[curr]
                    curr += 1
                if curr == length:
                    result.append(word)
        return result


print(Solution().findAndReplacePattern(["abc","deq","mee","aqq","dkd","ccc"], "abb"))

File: 2022/july/29-07-2022/find-and-replace-pattern.rs
use std::collections::HashMap;

impl Solution {
    pub fn find_and_replace_pattern(words: Vec<String>, pattern: String) -> Vec<String> {
        let p = compile(pattern.as_bytes());
        words.into_iter()
            .filter( |w| compile(w.as_bytes()) == p )
            .collect()
    }
}

fn compile(word: &[u8]) -> Vec<i32> {
    let N = word.len();
    let mut out = vec![ 0 as i32; N ];
    let mut perm: HashMap<u8, i32> = HashMap::new();
    let mut uniqueness: i32 = 0;
    for i in 0 .. N {
        let c = word[i];
        out[i] = *perm.entry(c)
            .or_insert_with( || {
                uniqueness += 1;
                uniqueness
            });
    }
    out
}